plugin/trino-snowflake/src/test/java/io/trino/plugin/snowflake/TestSnowflakeTypeMapping.java

```diff
@@ -0,0 +1,327 @@
+                .addRoundTrip("char(77)", "'Ну, погоди!'", createVarcharType(77), "CAST('Ну, погоди!' AS VARCHAR(77))")
+                .execute(getQueryRunner(), snowflakeCreateAndInsert("tpch.snowflake_test_parameterized_char"));
+    }
+
+    @Test
+    public void testBinary()
+    {
+        SqlDataTypeTest.create()
+                .addRoundTrip("binary(18)", "NULL", VARBINARY, "CAST(NULL AS varbinary)")
+                .addRoundTrip("binary(18)", "X''", VARBINARY, "X''")
+                .addRoundTrip("binary(18)", "X'68656C6C6F'", VARBINARY, "to_utf8('hello')")
+                .addRoundTrip("binary(18)", "X'C582C4856B61207720E69DB1E4BAACE983BD'", VARBINARY, "to_utf8('łąka w 東京都')") // no trailing zeros
+                .addRoundTrip("binary(18)", "X'4261672066756C6C206F6620F09F92B0'", VARBINARY, "to_utf8('Bag full of 💰')")
+                .addRoundTrip("binary(18)", "X'0001020304050607080DF9367AA7000000'", VARBINARY, "X'0001020304050607080DF9367AA7000000'") // non-text prefix
+                .addRoundTrip("binary(18)", "X'000000000000'", VARBINARY, "X'000000000000'")
+                .execute(getQueryRunner(), snowflakeCreateAndInsert("tpch.test_binary"));
+    }
+
+    @Test
+    public void testVarbinary()
+    {
+        SqlDataTypeTest.create()
+                .addRoundTrip("varbinary", "NULL", VARBINARY, "CAST(NULL AS varbinary)")
+                .addRoundTrip("varbinary", "X''", VARBINARY, "X''")
+                .addRoundTrip("varbinary", "X'68656C6C6F'", VARBINARY, "to_utf8('hello')")
+                .addRoundTrip("varbinary", "X'5069C4996B6E6120C582C4856B61207720E69DB1E4BAACE983BD'", VARBINARY, "to_utf8('Piękna łąka w 東京都')")
+                .addRoundTrip("varbinary", "X'4261672066756C6C206F6620F09F92B0'", VARBINARY, "to_utf8('Bag full of 💰')")
+                .addRoundTrip("varbinary", "X'0001020304050607080DF9367AA7000000'", VARBINARY, "X'0001020304050607080DF9367AA7000000'") // non-text
+                .addRoundTrip("varbinary", "X'000000000000'", VARBINARY, "X'000000000000'")
+                .execute(getQueryRunner(), trinoCreateAsSelect("test_varbinary"))
+                .execute(getQueryRunner(), snowflakeCreateAndInsert("tpch.test_varbinary"));
+    }
+
+    @Test(dataProvider = "sessionZonesDataProvider")
+    public void testDate(ZoneId sessionZone)
+    {
+        Session session = Session.builder(getSession())
+                .setTimeZoneKey(getTimeZoneKey(sessionZone.getId()))
+                .build();
+
+        SqlDataTypeTest.create()
+                .addRoundTrip("date", "NULL", DATE, "CAST(NULL AS DATE)")
+                .addRoundTrip("date", "DATE '0001-01-01'", DATE, "DATE '0001-01-01'") // Min value.
+                .addRoundTrip("date", "DATE '1582-10-05'", DATE, "DATE '1582-10-05'")
+                .addRoundTrip("date", "DATE '1582-10-14'", DATE, "DATE '1582-10-14'")
+                .addRoundTrip("date", "DATE '9999-12-31'", DATE, "DATE '9999-12-31'") // Max value.
```

Comment:
I believe the min and max are more large numbers.

Reply:
Here is a doc to explain the valid ranges of values. 
https://docs.snowflake.com/en/user-guide/date-time-input-output.html#valid-ranges-of-values-for-fields

According to his doc and my test, the ranges of date values to pass the statement of CAST('XXXX-XX-XX' as Date) is from '0001-01-01' to '9999-12-31'. So I think we can keep the current range in the test code.

You could check and confirm it. @ebyhr  😄

Reply:
I know the documentation is incorrect. 
Please add two test cases min - 1 & max + 1 so that we can verify what happens with such values and capture the future change in Snowflake. You can refer to `BaseSqlServerTypeMapping.testSqlServerDateUnsupported`

Reply:
To continue to track our issue and make it more clear, I pick our conversation which hasn't been resolved yet here from the Slack.

To sum up, tests cases of `0000-01-01, 9999-12-32, 99999-12-31` can't pass the SQL statement check with `Date` function in the system of snowflake, so there are the same problems while adding them in our tests of Trino project.

Snowflake can accept these values as `String` which accepts wide ranges comparing to date literal or function. So you suggested me to update my test cases of `Date` to `String`.

I tried 4 different transform methods to update Date test cases, but they did not give the expected results. You told me it was due to the date mapping implementation in `SnowflakeClient.java` is wrong with the case of `String`.

The current difficulty is that, in my current mapping columns function, I treat the Date as a type of `Long` and it works for these previous cases of Date. I don't well understand how to be compatible with both or how to figure out our current issue. Could you please give me some advice please ?

Reply:
I would recommend setting a break point where setting string literal in `SnowflakeClient.java` and confirm the actual value. The value should be difference from what you confirm in Snowflake UI.

Reply:
Yes, I referred to the function of `dateWriteFunctionUsingString()` in `PhoenixClient.java` and I debugged by setting the break point to the Type.Date. Finally, test cases passed the SQL statement check, but it always had a not correct expecting data.😭 After I explored it for some time, it was not solved. So I reverted my modification about this issue in order not to break all and added comments about the min and max of string in the test.

 Besides these two extremes, are there any other issue we need to explore, please? 😄

Reply:
Please fix the date formatter to `uuuu-MM-dd`.
```java
    private static final DateTimeFormatter DATE_FORMATTER = DateTimeFormatter.ofPattern("uuuu-MM-dd");
```

Reply:
Done and fixed the max and min test case. Thank you so much for your help. 😄 I think I can make this conversation as resolved.

Reply:
I believe the min and max value in the existing test are wrong.

Reply:
Hello, to confirm about this, I just run the tests of Date in `TestSnowflakeTypeMapping.java`, and they work well with my current commit. I'll send you the screenshot of running test.😄  What do you believe the min and max value are wrong, please ?
![image](https://user-images.githubusercontent.com/8734488/155718961-7a40f022-3565-427c-a7a7-aad7beeaf31d.png)

Reply:
Replied to your DM in Slack.

Reply:
Hello @ebyhr . I expanded the range of Date from `-5877641-06-23` to `5881580-07-11`, which are the min and max values in Trino and the tests passed. 😄